parsers/GB.py

```diff
@@ -1,71 +1,104 @@
-#!/usr/bin/env python3
-
-"""
-Parser that uses the RTE-FRANCE API to return the following data type(s)
-fetch_price method copied from FR parser.
-Day-ahead Price
-"""
+"""Parser that uses the RTE-FRANCE API"""
 
 import xml.etree.ElementTree as ET
-from datetime import datetime, timedelta
+from datetime import datetime, timedelta, timezone
 from logging import Logger, getLogger
+from zoneinfo import ZoneInfo
 
-import arrow
 from requests import Session
 
+from electricitymap.contrib.lib.models.event_lists import PriceList
+from electricitymap.contrib.lib.models.events import EventSourceType
+from electricitymap.contrib.lib.types import ZoneKey
 from parsers.lib.config import refetch_frequency
+from parsers.lib.exceptions import ParserException
+
+PARSER = "GB.py"
+TIMEZONE = ZoneInfo("Europe/London")
+ZONE_KEY = ZoneKey("GB")
 
 
-@refetch_frequency(timedelta(days=1))
+@refetch_frequency(timedelta(days=2))
 def fetch_price(
-    zone_key: str,
+    zone_key: ZoneKey = ZONE_KEY,
     session: Session | None = None,
     target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-) -> list:
-    if target_datetime:
-        now = arrow.get(target_datetime, tz="Europe/Paris")
-    else:
-        now = arrow.now(tz="Europe/London")
+) -> list[dict]:
+    """Returns one-hourly prices for the requested day (and the previous one)."""
 
-    r = session or Session()
-    formatted_from = now.shift(days=-1).format("DD/MM/YYYY")
-    formatted_to = now.format("DD/MM/YYYY")
+    now = datetime.now(timezone.utc)
+    target_datetime = (
+        now if target_datetime is None else target_datetime.astimezone(timezone.utc)
+    )
 
-    url = f"http://eco2mix.rte-france.com/curves/getDonneesMarche?dateDeb={formatted_from}&dateFin={formatted_to}&mode=NORM"
+    # API works in UTC timestamps
+    num_backlog_days = 1
+    day_start = (target_datetime - timedelta(days=num_backlog_days)).strftime(
+        "%d/%m/%Y"
+    )
+    day_end = target_datetime.strftime("%d/%m/%Y")
+    url = f"http://eco2mix.rte-france.com/curves/getDonneesMarche?dateDeb={day_start}&dateFin={day_end}&mode=NORM"
 
-    response = r.get(url)
-    obj = ET.fromstring(response.content)
-    datas = {}
+    session = session or Session()
+    response = session.get(url)
+
+    if not response.ok:
+        raise ParserException(
+            PARSER,
+            f"Exception when fetching price error code: {response.status_code}: {response.text}",
+            zone_key,
+        )
 
-    for donnesMarche in obj:
-        if donnesMarche.tag != "donneesMarche":
-            continue
+    xml_tree = ET.fromstring(response.content)
 
-        start_date = arrow.get(
-            arrow.get(donnesMarche.attrib["date"]).datetime, "Europe/Paris"
+    price_list = PriceList(logger=logger)
+    for daily_market_data in xml_tree.iterfind("donneesMarche"):
+        day = datetime.strptime(daily_market_data.get("date"), "%Y-%m-%d").replace(
```

Comment:
Small nitpick here, calling get on daily_market_data can potentially return none if the object is missing which is casuing type errors since strptime only accepts strings.